### PR TITLE
WIP: draft - compile and run src/mono/mono/tests tests

### PR DESCRIPTION
**Pull request by lambdageek**
Early draft PR just to get some feedback on the overall approach.

Overall approach:
1. Find all the tests in `src/mono/mono/tests` that are "simple" - a simple test is just a single assembly (perhaps with a dependency on `src/mono/mono/mini/TestDriver.cs`) that uses an exit code to signal success or failure.  It doesn't have a dependency on `libtest.c` and it doesn't require special compilation.
2. Write out a `.csproj` file somewhere under `artifacts/obj/` for each simple test
3. Compile the `.csproj` - it will put a test assembly in `artifacts/tests/mono/os.platform.config/testName/testName.dll`
4. Separate target that invokes each `.dotnet-mono/dotnet .../testName/testName.dll` one by one

---

TODO:
- [ ] Move the item group listing all the tests to a separate .props file so that it's mostly data and separate from the build logic
- [ ] Tests that need `libtest.{so,dylib}`
   - [ ] Add a test type for simple tests that depend on `libtest.c`
   - [ ] Add a target that compiles `libtest.c` (probably by invoking `make -C mono/tests compile-libtest` or something like that)
   - [ ] Add a step to the .csproj that copies the libtest native library to each test that needs it
- [ ] Tests that are comprised of multiple files or assemblies
   - [ ] Add a test type that uses its own `.projitems` and `.targets` files - use those for multi-file tests - the .projitems will have all the extra `Compile` items and the `.targets` can hook into the build for any funny business.  Not sure if this is enough for tests that need to build multiple assemblies
   - [ ] Modify the .csproj template to import the .projitems and .targets
- [ ] Add an xunit runner wrapper.  Probably can borrow a lot of coreclr's logic from their xunit wrapper in  `src/coreclr/tests/src/runtest.proj` .   Or maybe it's less effort to take our babysitter python script and teach it to emit an xunit-compatible output file.  And something about exclusions.
- [ ] Figure out AOT
- [ ] Figure out Helix and the Azure Pipelines

### CODE REVIEW

**Comment by lambdageek** [comment]
@akoeplinger does this seem like the right approach?  I'd like a sanity check before I take this too far

File: src/mono/mono-test.proj [hunk 1]
```diff
@@ -0,0 +1,135 @@
+<Project>
+  <Import Project="Directory.Build.props" />
+  <Import Project="Directory.Build.targets" />
```
Comment (jaredpar) [inline src/mono/mono-test.proj]: Why the explicit imports here? The import of these will be automatic with MSBuild. 

Also the targets file should generally speaking come at the end of the file vs. the beginning.

Comment (lambdageek) [reply src/mono/mono-test.proj]: Don't know. I'm copying from `src/mono/mono.proj` (which probably copied it from `src/runtime/runtime.proj` which does it this way).

I tried taking out those two lines and I see my `<Message>`s printing:
```
  Creating .csproj files for mono runtime tests
  Mono obj dir is mono//
  Mono obj tests dir is mono//mono/tests/
  There are 2 tests, 2 simple
    Csproj mono/mono/tests/array-12193/array-12193.csproj
    Csproj mono/mono/tests/debug-casts/debug-casts.csproj
```
which is wrong (the obj dirs shuld be somewhere under `artifacts`).  So the `<Import>`s do _something_ and taking them out seems to break things.

**Comment by marek-safar** [comment]
@ViktorHofer @directhex @akoeplinger could you please review the proposal?

File: src/mono/mono-test.proj (hunk 1 above)
Comment (akoeplinger) [reply src/mono/mono-test.proj]: The reason why we import the `Directory.Build.targets` at the top is that we need to override the default Restore/Build/Pack targets. It was indeed copied from CoreCLR, not sure why the automatic import isn't working.

File: src/mono/mono-test.proj [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  <!-- stolen from src/coreclr/tests/src/Directory.Build.props and modified -->
+  <PropertyGroup>
+    <BaseOutputPath>$(ArtifactsDir)\tests\mono</BaseOutputPath>
+    <BaseOutputPath Condition="'$(__TestRootDir)' != ''">$(__TestRootDir)</BaseOutputPath>
+    <BaseOutputPathWithConfig>$(BaseOutputPath)\$(PlatformConfigPathPart)\</BaseOutputPathWithConfig>
+    <BinDir>$(BaseOutputPathWithConfig)</BinDir>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <MonoTestSimpleCsprojTemplateTestDriver>$(MSBuildProjectDirectory)\mono\mini\TestDriver.cs</MonoTestSimpleCsprojTemplateTestDriver>
```
Comment (akoeplinger) [inline src/mono/mono-test.proj]: nit for consistency:
```suggestion
    <MonoTestSimpleCsprojTemplateTestDriver>$(MonoProjectRoot)\mono\mini\TestDriver.cs</MonoTestSimpleCsprojTemplateTestDriver>
```

**Review by akoeplinger — COMMENTED** [review]
It doesn't feel like the right approach to me and makes it hard to hook up to helix. We should be closer to what CoreCLR is doing, or rather move to a dotnet test setup like they are trying: https://github.com/dotnet/runtime/issues/33068

Unfortunately I can't point to a single example of how to do it right, hope the @dotnet/runtime-infrastructure folks have more insights :)

**Comment by lambdageek** [comment]
> It doesn't feel like the right approach to me and makes it hard to hook up to helix. We should be closer to what CoreCLR is doing, or rather move to a dotnet test setup like they are trying: #33068

@akoeplinger So there's two separate issues:
1. Running the tests.  Via `dotnet xunit.console.dll ...` - it makes sense that I should follow the CoreCLR lead for this.
2. Compiling the tests - at a high level there are two approaches, I think:
  1. Write a .csproj for every test we're interested in running
  2. Generate a .csproj for every test we're interested in running using a template (what I have in this PR).

Is my approach to compiling the tests also not workable?

**Pull request closed without merging**